db/db_impl/db_impl_files.cc

```diff
@@ -312,9 +312,25 @@ void DBImpl::FindObsoleteFiles(JobContext* job_context, bool force,
         // logs_ could have changed while we were waiting.
         continue;
       }
-      // This WAL file is not live, so it's OK if we never sync the rest of it
-      // or if we close it *after* removing from `logs_`. If it's already
-      // closed, then it's been fully synced.
+      // This WAL file is not live, so it's OK if we never sync the rest of it.
+      // If it's already closed, then it's been fully synced. If
+      // !background_close_inactive_wals then we need to Close it before
+      // removing from logs_ but not blocking while holding log_write_mutex_.
+      if (!immutable_db_options_.background_close_inactive_wals &&
+          log.writer->file()) {
+        // We are taking ownership of and pinning the front entry, so we can
+        // expect it to be the same after releasing and re-acquiring the lock
+        log.PrepareForSync();
+        log_write_mutex_.Unlock();
+        // TODO: maybe check the return value of Close.
+        // TODO: plumb Env::IOActivity, Env::IOPriority
+        auto s = log.writer->Close({});
```

Comment:
This seems to be a race with ` log.writer->file()` in `GetOpenWalSizes()`, I tried with TSAN and it reports so. The other `Close()` in `SyncWalImpl()` may be a race too.

Reply:
Because logs_ is a deque and we have pinned the first entry, it should not be a race. But that's too much for TSAN's brain to handle. I should be able to make it happy.

Reply:
> logs_ is a deque and we have pinned the first entry 

I think you are referring to the race between two calls to Close() for the same `log.writer`.
I meant the race between the read `log.writer->file()` and the write `log.writer->Close({})`.

Reply:
Ok, with the TSAN trace you sent it's more clearly an actual race. Nice catch!